src/dataflow/src/source/mod.rs

```diff
@@ -1676,11 +1676,23 @@ impl SourceReaderPersistence {
                 starting_offsets
                     .entry(source_timestamp.partition.clone())
                     .and_modify(|current_offset| {
-                        if source_timestamp.offset > *current_offset {
-                            *current_offset = source_timestamp.offset;
+                        match current_offset {
+                            Some(current_offset) if source_timestamp.offset > *current_offset => {
+                                *current_offset = source_timestamp.offset;
+                            }
+                            _ => (), // ignore "older" starting offsets
                         }
                     })
-                    .or_insert_with(|| source_timestamp.offset);
+                    .or_insert_with(|| Some(source_timestamp.offset));
+            } else {
```

Comment:
I _think_ it works. But it's definitely good to think about this again and again. Also, I'll sleep easier once we only have one "source" of persisted timestamps (either sqlite or persistence) again.

It does work because we always seed the `timestamp_histories` first with what we get from sqlite, then we potentially add more bindings that we have from persistence. In practice, it will not happen that we have bindings in persistence that we don't also have in sqlite, though.

In your example, we would still have the bindings `1 -> t1`, `2 -> t2` when restarting and go from that.

Reply:
ah ok. I was a little thrown off by this comment which seemed to imply the order is indeterminate (but i think its only btw timestamper and persist, and not the initial re-read from sqlite done in the coordinator)

```
 We need to do this because both restoring from persisted timestamp
        // bindings and the partition discovery thread running on the coordinator can lead to this
        // method being called, and the order is indeterminate.
```